cltk/phonology/old_english/orthophonology.py

```diff
@@ -0,0 +1,205 @@
+u_long = u.lengthen()
+e  = Vowel(Height.mid, Backness.front, Roundedness.neg, Length.short, 'e')
+e_long = e.lengthen()
+oe = Vowel(Height.mid, Backness.front, Roundedness.pos, Length.short, 'ø')
+oe_long = oe.lengthen()
+o  = Vowel(Height.mid, Backness.back, Roundedness.pos, Length.short, 'o')
+o_long = o.lengthen()
+ae = Vowel(Height.open, Backness.front, Roundedness.neg, Length.short, 'æ')
+ae_long = ae.lengthen()
+a  = Vowel(Height.open, Backness.back, Roundedness.neg, Length.short, 'ɑ')
+a_long = a.lengthen()
+
+# dipthongs
+aea = ae + a
+ae_long_a = ae_long + a
+eo = e + o
+e_long_o = e_long + o
+iu = i + u
+i_long_u = i_long + u
+
+
+consonants = [
+	m, n, n0, ng,
+	p, b, t, d, k, g,
+	tsh, dsh,
+	f, v, th, dh, s, z, sh, ch, x, y, h,
+	l, l0, j, w, w0,
+	r, r0
+]
+
+vowels = [
+	i, i_long, y, y_long, u, u_long,
+	e, e_long, oe, oe_long, o, o_long,
+	ae, ae_long, a, a_long
+]
+
+diphthongs = [
+	aea, ae_long_a, eo, e_long_o, iu, i_long_u
+]
+
+sound_inventory = consonants + vowels + diphthongs
+
+# these are the UNCONDITIONED sounds of the letters.
+alphabet = {
+	'a' : a,
+	'ā' : a_long,
+	'æ' : ae,
+	'ǣ' : ae_long,
+	'b' : b,
+	'c' : k,
+	'ċ' : tsh,
+	'd' : d,
+	'ð' : dh,
+	'e' : e,
+	'ē' : e_long,
+	'f' : f,
+	'g' : g,
+	'ġ' : j,
+	'h' : h,
+	'i' : i,
+	'ī' : i_long,
+	'l' : l,
+	'm' : m,
+	'n' : n,
+	'o' : o,
+	'ō' : o_long,
+	'p' : p,
+	'r' : r,
+	's' : s,
+	't' : t,
+	'u' : u,
+	'ū' : u_long,
+	'w' : w,
+	'x' : x,
+	'y' : y,
+	'ȳ' : y_long,
+	'þ' : th
+}
+
+diphthongs_ipa = {
+	'ea' : aea,
+	'ēa' : ae_long_a,
+	'ie' : i,
+	'īe' : i_long,
+	'eo' : eo,
+	'ēo' : e_long_o,
+	'io' : iu,
+	'īo' : i_long_u
+}
+
+digraphs_ipa = {
+	'cg' : dsh,
+	'ng' : ng,
+	'sc' : sh
+}
+
+oe = Orthophonology(sound_inventory, alphabet, diphthongs_ipa, digraphs_ipa)
+
+# intervocalic fricatives are voiced
+oe.add_rule(InnerPhonologicalRule(
+	condition = lambda before, target, after: 
+		before.is_vowel() and target in [f, s, th] and after.is_vowel(),
+	action = lambda target: 
+		oe.voice(target)))
+
+# /k/ is palatized in specific environments
+oe.add_rule(PhonologicalRule(
+	condition = lambda before, target, after:
+		target == k and
+		(after is not None and after[Backness] == Backness.front or
+		(before is not None and before == i and 
+			(after is None or (isinstance(after, Vowel) and after[Backness] != Backness.back)))),
+	action = lambda _: tsh))
+
+# palatization of /g/
+oe.add_rule(InnerPhonologicalRule(
+	condition = lambda before, target, after:
+		target == g and
+		(after[Backness] == Backness.front or
+		(before[Backness] == Backness.front and not(after[Backness] == Backness.back))),
+	action = lambda _ : j))
+
+# /g/ is fricativized when intervocalic
+oe.add_rule(InnerPhonologicalRule(
+	condition = lambda before, target, after:
+		(before.is_vowel() or before[Voiced] == Voiced.pos) and
+		target == g and
+		(after.is_vowel() or after[Voiced] == Voiced.pos),
+	action = lambda _ : y))

```

Comment:
What does this line do?

Reply:
It says: if the target is `g` and if the preceding segment and following segments are vocalized (vowels or [+voice] consonants) then emit `y`